src/hsm_cl.rs

```diff
@@ -0,0 +1,134 @@
+use class_group::primitives::cl_dl_lcm::{CLDLProofPublicSetup, Witness, HSMCL, PK};
+use curv::arithmetic::traits::Converter;
+use curv::arithmetic::traits::Samplable;
+use curv::elliptic::curves::traits::ECPoint;
+use curv::elliptic::curves::traits::ECScalar;
+use curv::BigInt;
+use curv::FE;
+use curv::GE;
+
+pub use class_group::primitives::cl_dl_lcm::Ciphertext;
+pub type PublicKey = PK;
+
+pub type Proof = CLDLProofPublicSetup;
+
+lazy_static::lazy_static! {
+     pub static ref PUBLIC_SETUP: BigInt = BigInt::from(b"A2LPoC".as_ref());
+}
+
+#[derive(Clone, Debug)]
+pub struct KeyPair(HSMCL);
+
+impl KeyPair {
+    pub fn public_key(&self) -> &PublicKey {
+        &self.0.pk
+    }
+
+    pub fn gen() -> Self {
+        Self(HSMCL::keygen_with_setup(&FE::q(), &1348, &*PUBLIC_SETUP))
+    }
+}
+
+pub fn encrypt(public_key: &PublicKey, message: &crate::secp256k1::KeyPair) -> (Ciphertext, Proof) {
+    let r = BigInt::sample_below(&(&public_key.stilde * BigInt::from(2).pow(40)));
+    let x = BigInt::from(message.secret_key().serialize().as_ref());
+    let ciphertext = HSMCL::encrypt_predefined_randomness(&public_key, &x, &r);
+    let pk_untagged_bytes = &message.public_key().serialize()[1..];
+    let X = GE::from_bytes(pk_untagged_bytes).unwrap();
+
+    let proof = CLDLProofPublicSetup::prove(
+        Witness { x: &x, r },
+        &public_key,
+        &ciphertext,
+        &X,
+        &*PUBLIC_SETUP,
+    );
+
+    (ciphertext, proof)
+}
+
+#[must_use]
+pub fn verify(
+    pk: &PublicKey,
+    ciphertext: &Ciphertext,
+    encrypts: &crate::secp256k1::PublicKey,
+    proof: &Proof,
+) -> bool {
+    let pk_untagged_bytes = &encrypts.serialize()[1..];
+    let encrypts = GE::from_bytes(pk_untagged_bytes).unwrap();
+    proof
+        .verify(pk, ciphertext, &encrypts, &*PUBLIC_SETUP)
+        .is_ok()
+}
+
+pub fn decrypt(keypair: &KeyPair, ciphertext: Ciphertext) -> secp256k1::curve::Scalar {
+    let bytes = BigInt::to_vec(&keypair.0.decrypt(&ciphertext));
+
+    // Note, if this isn't true then the problem should be solved at a lower level :^)
+    debug_assert!(
+        bytes.len() <= 32,
+        "decrypted value must < 32 because it was derived from discrete log of something of order q"
+    );
+
+    let mut bytes_32 = [0u8; 32];
+    // copy into the least significant bytes
+    bytes_32[32 - bytes.len()..].copy_from_slice(&bytes[..]);
+
+    let mut scalar = secp256k1::curve::Scalar::default();
+    let overflow: bool = scalar.set_b32(&bytes_32).into();
+    debug_assert!(
+        !overflow,
+        "this shouldn't overflow for the same reason as above"
+    );
+    scalar
+}
+
+pub fn multiply(ciphertext: &Ciphertext, sk: &secp256k1::SecretKey) -> Ciphertext {
```

Comment:
We also need the following function:

```rust
pub fn multiply(publickey: &secp256k1::PublicKey, sk: &secp256k1::SecretKey) -> secp256k1::PublicKey {
```

See line 22, figure 6, page 9 of the paper.

We thought that multiplication would require an instance of the public key but that doesn't seem to be the case so our (trait-based) model in `dummy_hsm_cl` is wrong. I think traits are still a good way of solving this though, but knowing that this is a globally available operation, I would probably implement `Mul` directly.

Reply:
I think we discussed this on call. it's just a usual ecmult and doesn't belong with hsm_cl. It is just a globally available operation but they didn't implement Mul for it yet.